server/src/main/java/org/opensearch/index/query/TemplateQueryBuilder.java

```diff
@@ -0,0 +1,186 @@
+        );
+    }
+
+    @Override
+    protected boolean doEquals(TemplateQueryBuilder other) {
+        return Objects.equals(this.content, other.content);
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(content);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    /**
+     * Gets the content of this template query.
+     *
+     * @return The template content as a map.
+     */
+    public Map<String, Object> getContent() {
+        return content;
+    }
+
+    /**
+     * Rewrites the template query by substituting variables from the context.
+     *
+     * @param queryCoordinatorContext The context for query rewriting.
+     * @return A rewritten QueryBuilder.
+     * @throws IOException If there's an error during rewriting.
+     */
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryCoordinatorContext) throws IOException {
+        // the queryRewrite is expected at QueryCoordinator level
+        if (!(queryCoordinatorContext instanceof QueryCoordinatorContext)) {
+            throw new IllegalStateException(
+                "Template Query must be rewritten at the coordinator node. Rewriting at shard level is not supported."
+            );
+        }
+
+        QueryCoordinatorContext convertedQueryCoordinateContext = (QueryCoordinatorContext) queryCoordinatorContext;
+        Map<String, Object> contextVariables = convertedQueryCoordinateContext.getContextVariables();
+        String queryString;
+
+        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+            builder.map(this.content);
+            queryString = builder.toString();
+        }
+
+        // Convert Map<String, Object> to Map<String, String> with proper JSON escaping
+        Map<String, String> variablesMap = contextVariables.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> {
+            try {
+                return JsonXContent.contentBuilder().value(entry.getValue()).toString();
+            } catch (IOException e) {
+                throw new RuntimeException("Error converting contextVariables to JSON string", e);
+            }
+        }));
+
+        String newQueryContent = replaceVariables(queryString, variablesMap);
+
+        try {
+            XContentParser parser = XContentType.JSON.xContent()
+                .createParser(queryCoordinatorContext.getXContentRegistry(), LoggingDeprecationHandler.INSTANCE, newQueryContent);
+
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+
+            QueryBuilder newQueryBuilder = parseInnerQueryBuilder(parser);
+
+            return newQueryBuilder;
+
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Failed to rewrite template query: " + newQueryContent, e);
+        }
+    }
+
+    private String replaceVariables(String template, Map<String, String> variables) {
```

Comment:
You should do an input sanity check and shortcut return here, for example:
 - null or empty `variables` map just return the template directly
 - null `template` return null or exception?
 - do an initial `indexOf()` check for lack of a substitution pattern.  
   - Basically, move your `startVar` on line 165 moved up).  Test if it's < 0 before going through the whole round trip of a new `StringBuilder`, appending the whole substring, and getting the string back out.
   - Potentially change `while (true) { .... break; }` to `do ... while(startVar >= 0) { ... recalculate startVar at end}`

Reply:
Another approach rather than iterating the string and trying to match start/end, is to iterate the `variables` map and construct search strings, e.g., for key `foo` create the string `${foo}` and then you can search for the entire string.   Check out [how Flow Framework](https://github.com/opensearch-project/flow-framework/blob/06b78310a1d786e1c69f16751177ecb843abf6fe/src/main/java/org/opensearch/flowframework/util/ParseUtils.java#L606-L643) does these substitutions. We use regex but could equally use an `indexOf`/`substring` approach.

Reply:
right,  I can check if the variables is null, then return the template directly. when null template then should throw exception, but this would ideally never happens, unless there is a mistake in the OpenSearch. I will make these change. 

some value in the map of variables of the map can be optional, that's why I am thinking we can accept null variables in the 

for example: 

```
GET _search
{
  "query": {
    "multi_match": {
      "query": "vector_search",
      "fields": ["${field1}","${field2}"]
    }
  }
}


```

Reply:
added null and empty checks in the latest commit [add empty or null check](https://github.com/opensearch-project/OpenSearch/pull/16818/commits/fcc44d52fec9c52746894a2dd8641e8d4c279c8d) also rebased with conflicts